internal/controllers/machineset/machineset_preflight.go

```diff
@@ -0,0 +1,220 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machineset
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	"sigs.k8s.io/cluster-api/feature"
+	"sigs.k8s.io/cluster-api/internal/contract"
+)
+
+type preflightCheckError *string
+
+// preflightFailedRequeueAfter is used to requeue the MachineSet to re-verify the preflight checks if
+// the preflight checks fail.
+const preflightFailedRequeueAfter = 15 * time.Second
+
+func (r *Reconciler) runPreflightChecks(ctx context.Context, cluster *clusterv1.Cluster, ms *clusterv1.MachineSet, action string) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	// If the MachineSetPreflightChecks feature gate is disabled return early.
+	if !feature.Gates.Enabled(feature.MachineSetPreflightChecks) {
+		return ctrl.Result{}, nil
+	}
+
+	skipped := skippedPreflightChecks(ms)
+	// If all the preflight checks are skipped then return early.
+	if skipped.Has(clusterv1.MachineSetPreflightCheckAll) {
+		return ctrl.Result{}, nil
+	}
+
+	// If the cluster does not have a control plane reference then there is nothing to do. Return early.
+	if cluster.Spec.ControlPlaneRef == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Get the control plane object.
+	controlPlane, err := external.Get(ctx, r.Client, cluster.Spec.ControlPlaneRef, cluster.Namespace)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q: failed to perform preflight checks: failed to get ControlPlane %s", action, klog.KRef(cluster.Spec.ControlPlaneRef.Namespace, cluster.Spec.ControlPlaneRef.Name))
+	}
+	cpKlogRef := klog.KRef(controlPlane.GetNamespace(), controlPlane.GetName())
+
+	// If the Control Plane version is not set then we are dealing with a control plane that does not support version
+	// or a control plane where the version is not set. In both cases we cannot perform any preflight checks as
+	// we do not have enough information. Return early.
+	cpVersion, err := contract.ControlPlane().Version().Get(controlPlane)
+	if err != nil {
+		if errors.Is(err, contract.ErrFieldNotFound) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q: failed to perform preflight checks: failed to get the version of ControlPlane %s", action, cpKlogRef)
```

Comment:
q: should this be treated as error or simply became a no-op given that preflight checks do not apply if CP do not support version?

Reply:
I think this case is handled in l.78. This here is the case where the CP has the field but it is not a string.

I would prefer returning the error here to surface the issue